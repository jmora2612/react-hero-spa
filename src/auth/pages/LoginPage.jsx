import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../context";

export const LoginPages = () => {
  //hook que se utiliza para navegar dentro de la app a alguna ruta
  const { login } = useContext(AuthContext);
  const navigate = useNavigate();

  const onLogin = () => {
    login("Jessi Mora");
    const lastPath = localStorage.getItem('lastPath')  || '/search'
    navigate(lastPath, { replace: true });
  };

  return (
    <div className="container mt-5">
      <h1>Login</h1>
      <hr />

      <button className="btn btn-primary" onClick={onLogin}>
        Login
      </button>
    </div>
  );
};
