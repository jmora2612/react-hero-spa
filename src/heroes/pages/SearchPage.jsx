import { useLocation, useNavigate } from 'react-router-dom';
import { useForm } from '../../hooks/UseForm';
import {HeroCard, HeroList} from '../components/index'
import queryString from 'query-string';
import { getHeroesByName } from '../helpers';

export const SearchPage = () => {
  const navigate = useNavigate();
  const location = useLocation();

  //queryString nos ayuda a procesar los query params en la ruta para extraer los datos 
  const {q = ''} = queryString.parse(location.search);
  const heroes = getHeroesByName(q);  
  const {searchText, onInputChange} =useForm({
    searchText:q
  })

  const alertDanger = <div aria-label='alert-danger' className="alert alert-danger animate__animated animate__fadeIn">Not hero with <b>{q}</b></div>
  const searchHero = <div className="alert alert-primary animate__animated animate__fadeIn">Search hero</div>

  const onSearchSubmit = (event)=>{
    event.preventDefault();
    navigate(`?q=${searchText}`)
    
  }

  return (
    <>
      <h1>SearchPage</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <h4>Searching</h4>
          <hr />
          <form onSubmit={onSearchSubmit} aria-label="form">
            <input
              type="text"
              placeholder="Search hero"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={onInputChange}
            />
            <button className="btn btn-outline-primary mt-2">Search</button>
          </form>
        </div>
        <div className="col-7">
          <h4>Result</h4>
          <hr />

          {(!q && searchHero)}
          {((heroes.length === 0 && q) && alertDanger)}
          { heroes.map((el) => (<HeroCard key={el.id} {...el}/>))}
        </div>
      </div>
    </>
  );
};
