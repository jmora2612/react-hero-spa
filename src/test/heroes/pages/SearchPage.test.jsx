import { fireEvent, render, screen } from "@testing-library/react"
import { MemoryRouter, useNavigate } from "react-router-dom"
import { SearchPage } from "../../../heroes/pages/SearchPage"


const mockedUserNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: () => mockedUserNavigate,
  }));
describe('prueba SearchPage', () => { 
    beforeEach(()=> jest.clearAllMocks())
    test('debe de mostrarse correctamente con valores por def', () => { 
        const {container}= render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        )

        expect(container).toMatchSnapshot()
     })

     test('debe de mostrar a batman y el input con el valor del queryString', () => { 
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        )

        const input = screen.getAllByRole('textbox')
        expect(input.value).toBe(batman);
        const img = screen.getAllByRole('img')
        expect(img.src).toContain('assets/heroes/dc-batman.jpg');
        const alertDanger = screen.getByLabelText("alert-danger");
        expect(alertDanger.style.display).tobe('none');
     }) 


     test('debe de mostrar un error si no se encuentra el hero', () => { 
        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchPage />
            </MemoryRouter>
        )
        const alertDanger = screen.getByLabelText("alert-danger");
        expect(alertDanger.style.display).toBe('');
      })

      test('debe de llamar el navigate a la pantalla nueva', () => { 
        const hero = 'superman';
        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        )

        const input = screen.getAllByRole('textbox')
        fireEvent.change(input, {target:{name:'searchText', value:hero}});
        const form = screen.getByRole('form');
        fireEvent.submit(form)
        expect(mockedUserNavigate).toHaveBeenCalledWith(`?q=${hero}`, {"replace": true});

      })
 })