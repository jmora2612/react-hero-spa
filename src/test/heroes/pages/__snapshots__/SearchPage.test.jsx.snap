// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prueba SearchPage debe de mostrarse correctamente con valores por def 1`] = `<div />`;
